src/Raven.Client/Util/TcpUtils.cs

```diff
@@ -241,51 +256,99 @@ private static TcpClient NewTcpClient(TimeSpan? timeout, bool useIPv6)
-                        tcpClient = await ConnectAsync(
-                            url,
-                            tcpConnectionTimeout
+                    tcpClient?.Dispose();
+                    stream?.Dispose();
+                    
+                    logs?.Add($"Trying to connect to :{url}");
+
+                    tcpClient = await ConnectAsync(
+                        url, 
+                        timeout
+#if TCP_CLIENT_CANCELLATIONTOKEN_SUPPORT
+                        ,
+                        token: token
+#endif
+                        ).ConfigureAwait(false);
+                    
+                    stream = await TcpUtils.WrapStreamWithSslAsync(
+                        tcpClient,
+                        info,
+                        cert,
+#if SSL_STREAM_CIPHERSUITESPOLICY_SUPPORT
+                        cipherSuitesPolicy,
+#endif
+                        timeout
 #if TCP_CLIENT_CANCELLATIONTOKEN_SUPPORT
-                            ,
-                            token: token
+                        ,
+                        token: token
 #endif
-                            ).ConfigureAwait(false);
-                        return (tcpClient, url);
+                    ).ConfigureAwait(false);
+
+                    switch (operationType)
+                    {
+                        case TcpConnectionHeaderMessage.OperationTypes.Subscription:
+                        case TcpConnectionHeaderMessage.OperationTypes.Replication:
+                        case TcpConnectionHeaderMessage.OperationTypes.Heartbeats:
+                        case TcpConnectionHeaderMessage.OperationTypes.Cluster:
+                            supportedFeatures = await negotiationCallback(url, info, stream, negContext).ConfigureAwait(false);
+                            break;
+                        case TcpConnectionHeaderMessage.OperationTypes.TestConnection:
+                        case TcpConnectionHeaderMessage.OperationTypes.Ping:
+                            supportedFeatures = await negotiationCallback(url, info, stream, negContext, logs).ConfigureAwait(false);
+                            break;
+                        default:
+                            throw new NotSupportedException($"Operation type '{operationType}' not supported.");
                     }
-                    catch
+
+                    logs?.Add($"{Environment.NewLine}Negotiation successful for operation {operationType}.{Environment.NewLine} {tcpClient.Client.LocalEndPoint} "+
+                              $"is connected to {tcpClient.Client.RemoteEndPoint}{Environment.NewLine}");
+
+                    return (tcpClient, stream, url, supportedFeatures);
+                }
+                catch (Exception e)
+                {
+                    logs?.Add($"Failed to connect to url {url}: {e.Message}");
+                    if (i == infoUrls.Length - 1)
                     {
-                        // ignored
+                        throw;
```

Comment:
Here, aggregate all the previous exceptions and throw an `AggregateException` if there is more than once.
We want to get the *full* errors for this scenario

Reply:
The problem is that I have to keep the type of the error for outer functions that rely on it. For example ShouldTryToReconnect in subscriptions.